graph/src/components/subgraph/proof_of_indexing/online.rs

```diff
@@ -0,0 +1,207 @@
+//! This is an online (streaming) implementation of the reference implementation
+//! Any hash constructed from here should be the same as if the same data was given
+//! to the reference implementation, but this is updated incrementally
+
+use super::ProofOfIndexingEvent;
+use crate::prelude::{debug, EthereumBlockPointer, Logger, SubgraphDeploymentId};
+use lazy_static::lazy_static;
+use stable_hash::crypto::{Blake3SeqNo, SetHasher};
+use stable_hash::prelude::*;
+use stable_hash::utils::AsBytes;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fmt;
+use web3::types::H256;
+
+lazy_static! {
+    static ref LOG_EVENTS: bool = std::env::var("GRAPH_LOG_POI_EVENTS")
+        .unwrap_or("false".into())
+        .parse::<bool>()
+        .expect("invalid GRAPH_LOG_POI_EVENTS");
+}
+
+pub struct BlockEventStream {
+    vec_length: usize,
+    seq_no: Blake3SeqNo,
+    digest: SetHasher,
+}
+
+// Performance: Could write a specialized function for this easily
+fn traverse_seq_no(counts: &[usize]) -> Blake3SeqNo {
+    counts.iter().fold(Blake3SeqNo::root(), |mut s, i| {
+        s.skip(*i);
+        s.next_child()
+    })
+}
+
+impl BlockEventStream {
+    fn new(block_number: u64) -> Self {
+        let events = traverse_seq_no(&[
+            1,                                // kvp -> v
+            0,                                // CausalityRegion.blocks: Vec<Block>
+            block_number.try_into().unwrap(), // Vec<Block> -> [i]
+            0,                                // Block.events -> Vec<ProofOfIndexingEvent>
+        ]);
```

Comment:
I'll step back a bit to discuss sequence numbers generally first.

Recall that `StableHash` does not write any data for fields that have a default value. It does not write the integer `0` for example. Without a `SequenceNumber`, writing data in this fashion creates trivial collisions in common cases.

A `SequenceNumber` uniquely identifies where some piece of data lives in the structure. In the code above, this expresses "The first member of the block_number'th member of the first member of the second member" (Read backward).  Members can be fields in a struct, elements of a `Vec`, etc.

The crypto stable hash is interesting in that it's just a huge multi-set hash. That means that any member can be added to the set in any order. Members of the set that are fed into this hash are all of the "SequenceNumber + Data" pairs.

So, this code and the code like it identifies where the data lives (using the reference implementation as canon) then calls `stable_hash` on the data with that sequence number.

In that way we build up a hash equivalent to the hash of the reference structure, but with one piece of data at a time.

Reply:
I think I understand. I'd find the following comment helpful (me):
```rust
// Obtain the sequence number that corresponds to the events in this stream.
// This sequence number is relative to a `(name, region)` tuple. The sequence
// traverses to the region in the tuple, the blocks in the region, the block with the
// block number that this stream is for, and then the events member in in that block.
let events = traverse_seq_no(&[
    1,                                // (name, region).1 — the causality region itself
    0,                                // region.blocks — blocks of the causality region (0th member of the region struct)
    block_number.try_into().unwrap(), // region.blocks[block_number] — the block the stream corresponds to
    0,                                // region.blocks[block_number].events — the events in the block (0th member of the block struct)
]);
```

Once you understand the sequence numbers, this may be obvious and we don't have to add comments like this to all `traverse_seq_no` calls, that would be excessive.

Reply:
Perhaps this could become a comment on `traverse_seq` that uses this particular example to illustrate how it works.

Reply:
I have pushed the comment with a complete example of a struct, data, and field to identify